include/boost/math/special_functions/detail/bernoulli_details.hpp

```diff
@@ -530,18 +496,35 @@ class bernoulli_numbers_cache
    atomic_counter_type m_counter, m_current_precision;
 };
 
-template <class T, class Policy>
-inline bernoulli_numbers_cache<T, Policy>& get_bernoulli_numbers_cache()
+template <class T, class Policy, int N>
+inline bernoulli_numbers_cache<T, Policy>& get_bernoulli_numbers_cache_imp(const std::integral_constant<int, N>&)
 {
    //
-   // Force this function to be called at program startup so all the static variables
-   // get initialized then (thread safety).
+   // Regular case, rely on C++11 thread safe initialization of the static variable:
    //
-   bernoulli_initializer<T, Policy>::force_instantiate();
    static bernoulli_numbers_cache<T, Policy> data;
    return data;
 }
 
+template <class T, class Policy>
+inline bernoulli_numbers_cache<T, Policy>& get_bernoulli_numbers_cache_imp(const std::integral_constant<int, 0>&)
+{
+   //
+   // If we get here, then the precision of T is unknown at compile time, which means
+   // there's a fair chance it's a variable precision type.  In that case we *must*
+   // make our cache thread local, that way we can have different threads operating
+   // at differing precisions:
+   //
+   static BOOST_MATH_THREAD_LOCAL bernoulli_numbers_cache<T, Policy> data;
```

Comment:
It makes no difference, but `BOOST_MATH_THREAD_LOCAL` always expands to `thread_local`. I should have axed that macro like the other pre-C++11 workarounds.

Reply:
Nod.  I did wonder about that, I guess I was worrying about platforms which might not have any thread support at all, but I guess those can just `#define thread_local` and be done with it.